UI/window-basic-filters.cpp

```diff
@@ -1225,6 +1225,55 @@ void OBSBasicFilters::PasteFilter()
 	OBSSource filter = OBSGetStrongRef(main->copyFilter);
 	if (!filter)
 		return;
+	char *new_name = obs_source_copy_single_filter2(source, filter);
+
+	const char *source_name = obs_source_get_name(source);
+
+	obs_data_t *undo_wrapper = obs_data_create();
+	obs_data_set_string(undo_wrapper, "source-name", source_name);
+	obs_data_set_string(undo_wrapper, "filter-name", new_name);
+	std::string undo_data = obs_data_get_json(undo_wrapper);
+	obs_data_release(undo_wrapper);
+	bfree(new_name);
+
+	obs_data_t *redo_wrapper = obs_data_create();
+	obs_data_set_string(redo_wrapper, "destination-name", source_name);
+	obs_data_set_string(redo_wrapper, "parent-name",
+			    obs_source_get_name(obs_filter_get_parent(filter)));
+	obs_data_set_string(redo_wrapper, "filter-name",
+			    obs_source_get_name(filter));
+	std::string redo_data = obs_data_get_json(redo_wrapper);
+	obs_data_release(redo_wrapper);
+
+	auto undo = [](const std::string &data) {
+		obs_data_t *undo_wrapper =
+			obs_data_create_from_json(data.c_str());
+		obs_source_t *source = obs_get_source_by_name(
+			obs_data_get_string(undo_wrapper, "source-name"));
+		obs_source_t *filter = obs_source_get_filter_by_name(
+			source,
+			obs_data_get_string(undo_wrapper, "filter-name"));
+		obs_data_release(undo_wrapper);
+
+		obs_source_filter_remove(source, filter);
+	};
+
+	auto redo = [](const std::string &data) {
```

Comment:
Same thing with the undo lambda, this does not release the references it adds to the sources.